src/mococrw/ecies.h

```diff
@@ -0,0 +1,332 @@
+     */
+    void setAuthTag( const std::vector<uint8_t>& tag);
+
+private:
+
+    ECIESDecryptionCtx(const AsymmetricPrivateKey& privKey,
+                       const std::shared_ptr<mococrw::KeyDerivationFunction> kdf,
+                       const std::shared_ptr<mococrw::MessageAuthenticationCode> mac,
+                       const mococrw::SymmetricCipherMode mode,
+                       const mococrw::SymmetricCipherKeySize keySize,
+                       const mococrw::SymmetricCipherPadding padding,
+                       const std::vector<uint8_t>& iv,
+                       const std::vector<uint8_t>& S1,
+                       const std::vector<uint8_t>& S2);
+
+    friend ECIESDecryptionCtxBuilder;
+
+    /**
+     * Internal class for applying the PIMPL design pattern
+     * (to hide the details of storing the padding objects from the client)
+     */
+    class Impl;
+
+    /**
+     * Pointer for PIMPL design pattern
+     */
+    std::unique_ptr<Impl> _impl;
+};
+
+/**
+ * @brief The ECIESCtxBuilder class is an pseudo abstract class for providing common functions to the Encryption and
+ * Decryption context builders.
+ */
+class ECIESCtxBuilder
+{
+public:
+    /**
+     * @brief Sets the key derivation function used for ECIES (Default: mococrw::X963KDF with SHA512)
+     * @param kdf The actual object used in ECIES
+     * @return builder instance
+     */
+    ECIESCtxBuilder &setKDF(std::shared_ptr<mococrw::KeyDerivationFunction> kdf);
+
+    /**
+     * @brief Sets the message authentication code used for ECIES. (Default: mococrw::HMAC with SHA512)
+     * @param mac The actual object used in ECIES
+     * @return builder instance
+     */
+    ECIESCtxBuilder &setMAC(std::shared_ptr<mococrw::MessageAuthenticationCode> mac);
+
+    /**
+     * @brief Sets the symmetric cipher mode. (Default: mococrw::SymmetricCipherMode::CBC)
+     * @param mode The mode used in ECIES
+     * @return builder instance
+     */
+    ECIESCtxBuilder &setSymmetricCipherMode(mococrw::SymmetricCipherMode mode);
+
+    /**
+     * @brief Sets the key size for the symmetric encryption. (Default: mococrw::SymmetricCipherKeySize::S_256)
+     * @param keySize The key size to be used
+     * @return builder instance
+     */
+    ECIESCtxBuilder &setSymmetricCipherKeySize(mococrw::SymmetricCipherKeySize keySize);
+
+    /**
+     * @brief Sets the padding for the symmetric encryption. (Default: mococrw::SymmetricCipherPadding::NO)
+     * @param padding The padding scheme to be used.
+     * @return builder instance
+     */
+    ECIESCtxBuilder &setSymmetircCipherPadding(mococrw::SymmetricCipherPadding padding);
+
+    /**
+     * @brief Sets the initialisation vector (IV) for the symmetric encryption. (Default: zeros)
+     * @param iv The initialisation vector. The length shall be the same as the key length
+     * @return builder instance
+     */
+    ECIESCtxBuilder &setSymmetricCipherIv(std::vector<uint8_t> iv);
```

Comment:
```suggestion
    ECIESCtxBuilder &setSymmetricCipherIv(const std::vector<uint8_t>& iv);
```
any reason to no use const-refs here (and in the other places)?

Reply:
Actually, if this is a copy, it could be `std::move()`d, and that may not be the worst idea.